// mat-contenteditable/esm5/lib/form-field-sizer.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Renderer2, ElementRef, AfterContentInit } from '@angular/core';
var FormFieldSizerDirective = /** @class */ (function () {
    function FormFieldSizerDirective(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
    }
    FormFieldSizerDirective.prototype.ngAfterContentInit = function () {
        this.updateSize();
    };
    FormFieldSizerDirective.prototype.updateSize = function () {
        var _this = this;
        var infix = this.getElement('mat-form-field-infix');
        this.renderer.removeStyle(infix, 'min-height');
        setTimeout(function () {
            var wrapper = _this.getElement('mat-form-field-wrapper');
            var offset = _this.elementRef.nativeElement.offsetHeight -
                wrapper.offsetHeight -
                parseFloat(getComputedStyle(wrapper).marginTop) -
                parseFloat(getComputedStyle(wrapper).marginBottom) +
                parseFloat(getComputedStyle(infix).height);
            _this.renderer.setStyle(infix, 'min-height', offset + "px");
        });
    };
    FormFieldSizerDirective.prototype.getElement = function (name) {
        return this.elementRef.nativeElement.getElementsByClassName(name).item(0);
    };
    FormFieldSizerDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    FormFieldSizerDirective = __decorate([
        Directive({
            selector: '[formFieldSizer]'
        }),
        __metadata("design:paramtypes", [Renderer2,
            ElementRef])
    ], FormFieldSizerDirective);
    return FormFieldSizerDirective;
}());
export { FormFieldSizerDirective };
