// mat-contenteditable/esm5/lib/mat-ckeditor-balloon.directive.js
import { __decorate, __extends, __metadata } from "tslib";
import { Directive, Input } from '@angular/core';
import { MatFormFieldControl } from '@angular/material/form-field';
import { MatCkeditorDirective } from './mat-ckeditor.directive';
var MatCkeditorBalloonDirective = /** @class */ (function (_super) {
    __extends(MatCkeditorBalloonDirective, _super);
    function MatCkeditorBalloonDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MatCkeditorBalloonDirective_1 = MatCkeditorBalloonDirective;
    Object.defineProperty(MatCkeditorBalloonDirective.prototype, "toolbar", {
        set: function (show) {
            if (this.editor && show !== this.toolbarOpen) {
                var balloon = this.editor.editorInstance.plugins.get('BalloonToolbar');
                if (show) {
                    this.showToolbar(balloon);
                }
                else {
                    balloon.hide();
                    this.toolbarOpen = false;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    MatCkeditorBalloonDirective.prototype.ngOnInit = function () {
        var _this = this;
        _super.prototype.ngOnInit.call(this);
        this.editor.ready.subscribe(function (editor) {
            var balloon = editor.plugins.get('BalloonToolbar');
            balloon.stopListening(editor.model.document.selection, 'change:range');
            balloon.stopListening(balloon, '_selectionChangeDebounced');
        });
        this.editor.focus.subscribe(function () {
            if (_this.toolbarOpen) {
                var balloon = _this.editor.editorInstance.plugins.get('BalloonToolbar');
                _this.showToolbar(balloon);
            }
        });
    };
    MatCkeditorBalloonDirective.prototype.showToolbar = function (balloon) {
        if (!balloon._balloon.hasView(balloon.toolbarView)) {
            balloon.listenTo(this.editor.editorInstance.ui, 'update', function () {
                balloon._balloon.updatePosition(balloon._getBalloonPositionData());
            });
            balloon._balloon.add({
                view: balloon.toolbarView,
                position: balloon._getBalloonPositionData(),
                balloonClassName: 'ck-toolbar-container'
            });
            this.toolbarOpen = true;
        }
    };
    var MatCkeditorBalloonDirective_1;
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MatCkeditorBalloonDirective.prototype, "toolbar", null);
    MatCkeditorBalloonDirective = MatCkeditorBalloonDirective_1 = __decorate([
        Directive({
            selector: '[matCkeditorBalloon]',
            providers: [
                { provide: MatFormFieldControl, useExisting: MatCkeditorBalloonDirective_1 },
            ]
        })
    ], MatCkeditorBalloonDirective);
    return MatCkeditorBalloonDirective;
}(MatCkeditorDirective));
export { MatCkeditorBalloonDirective };
