// mat-contenteditable/esm2015/lib/form-field-sizer.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Renderer2, ElementRef, AfterContentInit } from '@angular/core';
let FormFieldSizerDirective = class FormFieldSizerDirective {
    constructor(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
    }
    ngAfterContentInit() {
        this.updateSize();
    }
    updateSize() {
        const infix = this.getElement('mat-form-field-infix');
        this.renderer.removeStyle(infix, 'min-height');
        setTimeout(() => {
            const wrapper = this.getElement('mat-form-field-wrapper');
            const offset = this.elementRef.nativeElement.offsetHeight -
                wrapper.offsetHeight -
                parseFloat(getComputedStyle(wrapper).marginTop) -
                parseFloat(getComputedStyle(wrapper).marginBottom) +
                parseFloat(getComputedStyle(infix).height);
            this.renderer.setStyle(infix, 'min-height', `${offset}px`);
        });
    }
    getElement(name) {
        return this.elementRef.nativeElement.getElementsByClassName(name).item(0);
    }
};
FormFieldSizerDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
FormFieldSizerDirective = __decorate([
    Directive({
        selector: '[formFieldSizer]'
    }),
    __metadata("design:paramtypes", [Renderer2,
        ElementRef])
], FormFieldSizerDirective);
export { FormFieldSizerDirective };
