// mat-contenteditable/esm2015/lib/mat-ckeditor-balloon.directive.js
var MatCkeditorBalloonDirective_1;
import { __decorate, __metadata } from "tslib";
import { Directive, Input } from '@angular/core';
import { MatFormFieldControl } from '@angular/material/form-field';
import { MatCkeditorDirective } from './mat-ckeditor.directive';
let MatCkeditorBalloonDirective = MatCkeditorBalloonDirective_1 = class MatCkeditorBalloonDirective extends MatCkeditorDirective {
    set toolbar(show) {
        if (this.editor && show !== this.toolbarOpen) {
            const balloon = this.editor.editorInstance.plugins.get('BalloonToolbar');
            if (show) {
                this.showToolbar(balloon);
            }
            else {
                balloon.hide();
                this.toolbarOpen = false;
            }
        }
    }
    ngOnInit() {
        super.ngOnInit();
        this.editor.ready.subscribe(editor => {
            const balloon = editor.plugins.get('BalloonToolbar');
            balloon.stopListening(editor.model.document.selection, 'change:range');
            balloon.stopListening(balloon, '_selectionChangeDebounced');
        });
        this.editor.focus.subscribe(() => {
            if (this.toolbarOpen) {
                const balloon = this.editor.editorInstance.plugins.get('BalloonToolbar');
                this.showToolbar(balloon);
            }
        });
    }
    showToolbar(balloon) {
        if (!balloon._balloon.hasView(balloon.toolbarView)) {
            balloon.listenTo(this.editor.editorInstance.ui, 'update', () => {
                balloon._balloon.updatePosition(balloon._getBalloonPositionData());
            });
            balloon._balloon.add({
                view: balloon.toolbarView,
                position: balloon._getBalloonPositionData(),
                balloonClassName: 'ck-toolbar-container'
            });
            this.toolbarOpen = true;
        }
    }
};
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], MatCkeditorBalloonDirective.prototype, "toolbar", null);
MatCkeditorBalloonDirective = MatCkeditorBalloonDirective_1 = __decorate([
    Directive({
        selector: '[matCkeditorBalloon]',
        providers: [
            { provide: MatFormFieldControl, useExisting: MatCkeditorBalloonDirective_1 },
        ]
    })
], MatCkeditorBalloonDirective);
export { MatCkeditorBalloonDirective };
